whylog/assistant/ranges.py

```diff
@@ -0,0 +1,53 @@
+"""
+Auxiliary methods on ranges (intervals)
+"""
+
+import six
+
+
+def complementary_ranges(ranges, start_index, end_index):
+    """
+    Returns complementary ranges from range (start_index, end_index)
+    For start_index = 0, end_index = 20 goes as follows:
+    ranges -> complementary_ranges
+    [(1, 5), (10, 15)] -> [(0, 1), (5, 10), (15, 20)]
+    [(1, 5), (2, 5)] -> [(0, 1), (5, 20)]
+    [(1, 5), (2, 10)] -> [(0, 1), (10, 20)]
+    """
+    ranges_union = _ranges_union(ranges)
+    complement_ranges = []
+
+    # edge cases
+    if not ranges:
+        return [(start_index, end_index)]
+    first_start, _ = ranges_union[0]
+    if start_index < first_start:
+        complement_ranges.append((start_index, first_start))
+    _, last_end = ranges_union[-1]
+    if end_index > last_end:
```

Comment:
if this `if` is moved behind the `for` loop, sorting at the end becomes unnecessary (?)